File: Clone_Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    /**
     * @param node: A undirected graph node
     * @return: A undirected graph node
     */
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        // write your code here
        unordered_map<const UndirectedGraphNode *, UndirectedGraphNode *> map;
        // return doCloneDFS(node, map);
        if (!node)
        {
            return nullptr;
        }
        doCloneBFS(node, map);
        return map[node];
    }
    void doCloneBFS(UndirectedGraphNode *node, unordered_map<const UndirectedGraphNode *, UndirectedGraphNode *> &map)
    {
        queue<const UndirectedGraphNode *> que;
        que.emplace(node);
        map[node] = new UndirectedGraphNode(node->label);
        while (!que.empty())
        {
            auto old_node = que.front();
            que.pop();
            for (const auto &n : old_node->neighbors)
            {
                if (map.find(n) == map.end())
                {
                    map[n] = new UndirectedGraphNode(n->label);
                    que.emplace(n);
                }
                map[old_node]->neighbors.emplace_back(map[n]);
            }
        }
        
    }
    
    UndirectedGraphNode *doCloneDFS(UndirectedGraphNode *node, unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> &map)
    {
        if (!node)
        {
            return nullptr;
        }
        if (map.count(node) == 0)
        {
            auto new_node = new UndirectedGraphNode(node->label);
            map[node] = new_node;
            for (auto &n : node->neighbors)
            {
                new_node->neighbors.emplace_back(doCloneDFS(n, map));
            }
            return new_node;
        }
        else
        {
            return map[node];
        }
    }
};

File: Clone Graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []
class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def __init__(self):
        self.dict = {}
        
    def cloneGraph(self, node):
        # write your code here
        # if not node:
        #     return None
        # queue = [node]
        # visited = set()
        # while queue:
        #     n = queue.pop(0)
        #     if n in visited:
        #         continue
        #     else:
        #         visited.add(n)
        #     if n not in self.dict:
        #         cpn = UndirectedGraphNode(n.label)
        #         self.dict[n] = cpn
        #     cpn = self.dict[n]
        #     for nb in n.neighbors:
        #         if nb not in self.dict:
        #             cpnb = UndirectedGraphNode(nb.label)
        #             self.dict[nb] = cpnb
        #         queue.append(nb)
        #         cpn.neighbors.append(self.dict[nb])
        # return self.dict[node]
        
        ## copy node
        if not node:
            return None
        queue = [node]
        while queue:
            n = queue.pop(0)
            ## version 1
            if n not in self.dict:
                self.dict[n] = UndirectedGraphNode(n.label)
                for nb in n.neighbors:
                    queue.append(nb)
            ## version 2
            # if n in self.dict:
            #     continue
            # self.dict[n] = UndirectedGraphNode(n.label)
            # for nb in n.neighbors:
            #     queue.append(nb)
        
        ## assign neighbors
        for n in self.dict.keys():
            cpn = self.dict[n]
            for nb in n.neighbors:
                cpn.neighbors.append(self.dict[nb])
        return self.dict[node]
        
        